core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,695 @@
+        else:
+            # Else mark the ID fo program so that it will be
+            # trained using SVM later.
+            missclassified_points.append(pid)
+
+    # Calculate average of occurrence. This is used during prediction so
+    # that if winner of nearest neighbours appears less than the occurrence
+    # we consider that KNN has failed in prediction.
+    occurrence /= float(len(training_data))
+
+    return occurrence, missclassified_points
+
+
+class CodeClassifier(base.BaseClassifier):
+    """A class for code classifier that uses supervised learning to match
+    Python programs to an answer group. The classifier trains on programs
+    that exploration editors have assigned to an answer group.
+    """
+    def __init__(self):
+        super(CodeClassifier, self).__init__()
+        self.training_data = None
+        self.clf = None
+        self.token_to_id = None
+        self.T = None
+        self.K = None
+        self.occurrence = None
+        self.class_to_answer_group_mapping = None
+        self.top = None
+        self.count_vector = None
+
+    def to_dict(self):
+        """Returns a dict representing this classifier.
+
+        Returns:
+            dict. A dictionary representation of classifier referred as
+            'classifier_data'. This data is used for prediction.
+        """
+        fingerprint_data = {
+            i: {
+                'fingerprint': self.training_data[i]['fingerprint'],
+                'class': self.training_data[i]['class']
+            } for i in self.training_data
+        }
+
+        classifier_data = {
+            'KNN': {
+                'token_to_id': self.token_to_id,
+                'T': self.T,
+                'K': self.K,
+                'top': self.top,
+                'occurrence': self.occurrence,
+                'fingerprint_data': fingerprint_data
+            },
+            'SVM': self.clf.__dict__,
+            'class_to_answer_group_mapping': self.class_to_answer_group_mapping,
+            'cv_vocabulary': self.count_vector.__dict__['vocabulary_']
+        }
+        return classifier_data
+
+    # pylint: disable=too-many-locals
+    def train(self, training_data):
+        """Trains classifier using given training_data.
+
+        Args:
+            training_data: list(dict). The training data that is used for
+                training the classifier. The list contains dicts where each dict
+                represents a single training data group, for example:
+                training_data = [
+                    {
+                        'answer_group_index': 1,
+                        'answers': ['a1', 'a2']
+                    },
+                    {
+                        'answer_group_index': 2,
+                        'answers': ['a2', 'a3']
+                    }
+                ]
+        """
+        # Answer group index to classifier class mapping.
+        answer_group_to_class_mapping = {
```

Comment:
I just thought it would be better to use different classes because answer groups can be [3, 5, 10]. So traning data is present only for 3rd, 5th and 10th answer group. In that case would not it be better to use [0, 1, 2] and map them to [3, 5, 10]? But I will change it to use answer group. (initially I thought it should not be possible to use answer group index and therefore I used mapping approach but now I think it should be possible.)